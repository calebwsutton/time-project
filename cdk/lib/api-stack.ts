import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as route53 from '@aws-cdk/aws-route53';
import * as route53targets from '@aws-cdk/aws-route53-targets';

export class ApiStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, domain: string, hz: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const subdomain = 'time-api';

    const hostedZone = route53.HostedZone.fromHostedZoneAttributes(this, "HostedZone", {
      hostedZoneId: hz,
      zoneName: domain
    });

    const certificate = new acm.DnsValidatedCertificate(this, 'ApiCertificate', {
      hostedZone: hostedZone,
      domainName: `${subdomain}.${domain}`
    });

    const func = new lambda.Function(this, 'LambdaHandler', {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: lambda.Code.fromAsset(`${__dirname}/../../api/build/`),
      handler: 'app.handler',
    });

    const restApi = new apigateway.LambdaRestApi(this, 'Api', {
      handler: func,
      domainName: {
        certificate: certificate,
        domainName: `${subdomain}.${domain}`
      }
    }); 

    new route53.ARecord(this, 'ApiARecord', {
      zone: hostedZone,
      target: route53.RecordTarget.fromAlias(new route53targets.ApiGatewayDomain(restApi.domainName!)),
      recordName: `${subdomain}.${domain}`
    });

  }
}
