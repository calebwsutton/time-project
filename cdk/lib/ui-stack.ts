import * as cdk from '@aws-cdk/core';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Bucket, BlockPublicAccess } from '@aws-cdk/aws-s3';
import {CloudFrontWebDistribution, ViewerCertificate, OriginAccessIdentity} from '@aws-cdk/aws-cloudfront';
import { DnsValidatedCertificate } from '@aws-cdk/aws-certificatemanager';
import * as route53 from '@aws-cdk/aws-route53';
import * as route53targets from '@aws-cdk/aws-route53-targets';

export class UiStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    
    const hostedZone = route53.HostedZone.fromHostedZoneAttributes(this, "HostedZone", {
      hostedZoneId: "Z09770092VPEX62M3LDPL",
      zoneName: "calebwsutton.com"
    });

    const certificate = new DnsValidatedCertificate(this, 'UiCertificate', {
      hostedZone: hostedZone,
      domainName: "time.calebwsutton.com"
    });

    const bucket = new Bucket(this, 'Bucket', {
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL
    });

    const oia = new OriginAccessIdentity(this, 'OIA');
    bucket.grantRead(oia);

    new BucketDeployment(this, 'DeploySite', {
      sources: [Source.asset(`${__dirname}/../../ui/dist`)],
      destinationBucket: bucket
    });

    const cfDistribution = new CloudFrontWebDistribution(this, 'CFDistribution', {
      originConfigs: [{
        s3OriginSource: {
          s3BucketSource: bucket,
          originAccessIdentity: oia
        },
        behaviors: [{isDefaultBehavior: true}]
      }],
      viewerCertificate: ViewerCertificate.fromAcmCertificate(certificate, {
        aliases: ["time.calebwsutton.com"]
      })
    });

    new route53.ARecord(this, 'ApiARecord', {
      zone: hostedZone,
      target: route53.RecordTarget.fromAlias(new route53targets.CloudFrontTarget(cfDistribution)),
      recordName: "time.calebwsutton.com"
    });

  }
}
