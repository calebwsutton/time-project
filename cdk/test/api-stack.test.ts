import { expect as expectCDK, haveResource, haveResourceLike, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as api from '../lib/api-stack';

const stack = new api.ApiStack(new cdk.App(), 'ApiTestStack');

test('creates lambda function', () => {
  expectCDK(stack).to(haveResourceLike("AWS::Lambda::Function",));
});

test('creates rest api', () => {
  expectCDK(stack).to(haveResource("AWS::ApiGateway::RestApi"));
});

test('creates acm certificate', () => {
  expectCDK(stack).to(haveResource("AWS::CloudFormation::CustomResource", {
    DomainName: "time-api.calebwsutton.com",
    HostedZoneId: "Z09770092VPEX62M3LDPL"
  }));
});

test('creates route 53 A record', () => {
  expectCDK(stack).to(haveResource("AWS::Route53::RecordSet", {
    Type: "A"
  }));
});
