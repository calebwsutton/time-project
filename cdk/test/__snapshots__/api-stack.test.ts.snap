// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cfn matches snapshot 1`] = `
Object {
  "Outputs": Object {
    "ApiEndpoint4F160690": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiF70053CD",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiDeploymentStageprod3EB9684E",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4ArtifactHash5E28809B": Object {
      "Description": "Artifact hash for asset \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3Bucket5BEBDCBE": Object {
      "Description": "S3 bucket for asset \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC": Object {
      "Description": "S3 key for asset version \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParametersa708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0ArtifactHashC7D958C1": Object {
      "Description": "Artifact hash for asset \\"a708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0\\"",
      "Type": "String",
    },
    "AssetParametersa708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0S3BucketAF2DFF6F": Object {
      "Description": "S3 bucket for asset \\"a708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0\\"",
      "Type": "String",
    },
    "AssetParametersa708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0S3VersionKey7C3BDDBC": Object {
      "Description": "S3 key for asset version \\"a708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ApiANY40671C4F": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "LambdaHandler212865DC",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "ApiF70053CD",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiANYApiPermissionApiTestStackApi12FEB067ANYE186B321": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LambdaHandler212865DC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiF70053CD",
              },
              "/",
              Object {
                "Ref": "ApiDeploymentStageprod3EB9684E",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiANYApiPermissionTestApiTestStackApi12FEB067ANY35404C1A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LambdaHandler212865DC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiF70053CD",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiARecordFA826994": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "ApiCustomDomain249F3BF3",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "ApiCustomDomain249F3BF3",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "Z09770092VPEX62M3LDPL",
        "Name": "time-api.calebwsutton.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "ApiAccountA18C9B29": Object {
      "DependsOn": Array [
        "ApiF70053CD",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "ApiCloudWatchRole73EC6FC4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "ApiCertificateCertificateRequestorFunction35B7F4E7": Object {
      "DependsOn": Array [
        "ApiCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy37D02D57",
        "ApiCertificateCertificateRequestorFunctionServiceRole86B3A678",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3Bucket5BEBDCBE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ApiCertificateCertificateRequestorFunctionServiceRole86B3A678",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApiCertificateCertificateRequestorFunctionServiceRole86B3A678": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy37D02D57": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:AddTagsToCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/Z09770092VPEX62M3LDPL",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApiCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy37D02D57",
        "Roles": Array [
          Object {
            "Ref": "ApiCertificateCertificateRequestorFunctionServiceRole86B3A678",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApiCertificateCertificateRequestorResource57CB3F69": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "time-api.calebwsutton.com",
        "HostedZoneId": "Z09770092VPEX62M3LDPL",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "ApiCertificateCertificateRequestorFunction35B7F4E7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "ApiCloudWatchRole73EC6FC4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiCustomDomain249F3BF3": Object {
      "Properties": Object {
        "DomainName": "time-api.calebwsutton.com",
        "EndpointConfiguration": Object {
          "Types": Array [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": Object {
          "Fn::GetAtt": Array [
            "ApiCertificateCertificateRequestorResource57CB3F69",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "ApiCustomDomainMapApiTestStackApi12FEB067C305E020": Object {
      "Properties": Object {
        "DomainName": Object {
          "Ref": "ApiCustomDomain249F3BF3",
        },
        "RestApiId": Object {
          "Ref": "ApiF70053CD",
        },
        "Stage": Object {
          "Ref": "ApiDeploymentStageprod3EB9684E",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "ApiDeploymentB17BE62D1249d2b7136f9a091cf3ad603c6fdf66": Object {
      "DependsOn": Array [
        "ApiproxyANY4946669B",
        "Apiproxy05443892",
        "ApiANY40671C4F",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiDeploymentStageprod3EB9684E": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "ApiDeploymentB17BE62D1249d2b7136f9a091cf3ad603c6fdf66",
        },
        "RestApiId": Object {
          "Ref": "ApiF70053CD",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiF70053CD": Object {
      "Properties": Object {
        "Name": "Api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "Apiproxy05443892": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "ApiF70053CD",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiproxyANY4946669B": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "LambdaHandler212865DC",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "Apiproxy05443892",
        },
        "RestApiId": Object {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiproxyANYApiPermissionApiTestStackApi12FEB067ANYproxy05E2CF8C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LambdaHandler212865DC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiF70053CD",
              },
              "/",
              Object {
                "Ref": "ApiDeploymentStageprod3EB9684E",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiproxyANYApiPermissionTestApiTestStackApi12FEB067ANYproxy18260CC8": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LambdaHandler212865DC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiF70053CD",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LambdaHandler212865DC": Object {
      "DependsOn": Array [
        "LambdaHandlerServiceRole8F7B4955",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersa708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0S3BucketAF2DFF6F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0S3VersionKey7C3BDDBC",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa708a2a86ba300fec8edb610061784430adc211cdee8858e3aa93124f44065b0S3VersionKey7C3BDDBC",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "app.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaHandlerServiceRole8F7B4955",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaHandlerServiceRole8F7B4955": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
